pkg/sql/set_cluster_setting.go

```diff
@@ -775,3 +793,68 @@ func toSettingString(
 		return "", errors.Errorf("unsupported setting type %T", setting)
 	}
 }
+
+// unsafeSettingInterlockInfo contains information about the current
+// session that is used by the unsafe setting interlock system.
+type unsafeSettingInterlockInfo struct {
+	sessionID    clusterunique.ID
+	setNonce     func(context.Context, int32) error
+	interlockKey string
+	nonce        int32
+}
+
+func (p *planner) makeUnsafeSettingInterlockInfo() unsafeSettingInterlockInfo {
+	return unsafeSettingInterlockInfo{
+		sessionID:    p.ExtendedEvalContext().SessionID,
+		interlockKey: p.SessionData().UnsafeSettingInterlockKey,
+		nonce:        p.SessionData().UnsafeSettingInterlockNonce,
+		setNonce: func(ctx context.Context, nonce int32) error {
+			return p.applyOnSessionDataMutators(ctx, false /* local */, func(m sessionDataMutator) error {
+				m.SetUnsafeSettingInterlockNonce(nonce)
+				return nil
+			})
+		},
+	}
+}
+
+const interlockKeySessionVarName = "unsafe_setting_interlock_key"
+
+// unsafeSettingInterlock ensures that changes to unsafe settings are
+// doubly confirmed by the operator by a special value in a session
+// variable.
+func unsafeSettingInterlock(
+	ctx context.Context,
+	st *cluster.Settings,
+	setting settings.Setting,
+	encodedValue string,
+	info unsafeSettingInterlockInfo,
+) error {
+	// The interlock key is a combination of:
+	// - the session ID, so that different sessions need different keys.
+	// - a nonce different per SET CLUSTER SETTING statement, so that
```

Comment:
personally, I think we could do without the nonce; we've already made you read an error saying you're in the danger zone and ack that you read that warning by supplying us with a token included the warning.

This feels like reasonable safeguards against accidental misuse, but forcing you to do it again in the same session for the same setting probably doesn't really add much additional safety, but does add additional annoyance. 

I think we'd be just as safe without the nonce, just using the session ID+ setting key, so that once you've enabled manipulation of a given setting key in a given session, you can re-adjust/tune without being blocked each time. Especially if a change goes badly, being able to quickly -- without an _additional_ warning round-trip -- put it back to a previous value might be nice.

Reply:
> Especially if a change goes badly, being able to quickly -- without an additional warning round-trip -- put it back to a previous value might be nice.

I have thought about this more (and tried it both ways), and I don't see this argument work in practice. The SQL statement to run (SET ... = ...) is printed right there in the error message and trivial to copy-paste. So I don't personally see this annoyance materialize or this cost a noticeable overhead in productivity.

What the nonce does however is a _mistaken_ repeated copy-paste of an old value after setting a new value, e.g. by recalling a statement from the SQL shell history. Which is a more realistic failure mode.

Reply:
I still think a single "key" per session or perhaps session+setting is plenty to ensure the user is aware they're going into the unsafe realm -- and far more than we have now -- and we could always come back add more complexity to do the per-set keys later if we see unintended repeat sets actually being a problem. I think a once-per-session "let me do bad stuff" is maybe a little easier to explain/incorporate into automation of bad-stuff testing, and would have slightly less complex code (less state) but don't feel strongly enough about either to hold this up further if you're unconvinced.

Reply:
i'm ok with doing it the way you describe. i'll try this next. agree it's already better.